pkg/agent/discoverycache.go

```diff
@@ -186,13 +187,37 @@ func validateAPIOnDataplane(consumerInstances []apiV1.ResourceInstance) {
 		// - externalAPIID should not be empty
 		// - externalAPIStage could be empty for dataplanes that do not support it
 		if externalAPIID != "" && !agent.apiValidator(externalAPIID, externalAPIStage) {
-			log.Infof("API deleted from dataplane, deleting the catalog item %s from AMPLIFY Central", consumerInstance.Title)
-			err := agent.apicClient.DeleteConsumerInstance(consumerInstance.Name)
-			if err != nil {
-				log.Errorf("Unable to delete catalog item %s from AMPLIFY Central, %s", consumerInstance.Title, err.Error())
-			} else {
-				log.Infof("Deleted catalog item %s from AMPLIFY Central", consumerInstance.Title)
-			}
+			deleteConsumerInstanceOrService(consumerInstance, externalAPIID, externalAPIStage)
+		}
+	}
+}
+
+func shouldDeleteService(apiID, stage string) bool {
+	// no agent-specific validator means to delete the service
+	if agent.deleteServiceValidator == nil {
+		return true
+	}
+	// let the agent decide if service should be deleted
+	return agent.deleteServiceValidator(apiID, stage)
+}
+
+func deleteConsumerInstanceOrService(consumerInstance *v1alpha1.ConsumerInstance, externalAPIID, externalAPIStage string) {
+	if shouldDeleteService(externalAPIID, externalAPIStage) {
+		log.Infof("API no longer exists on the dataplane; deleting the API Service and corresponding catalog item %s from AMPLIFY Central", consumerInstance.Title)
```

Comment:
Amplify?

Reply:
done

Reply:
you get the other places in that method as well?